pkg/operator/csidriveroperator/olmremovalcontroller.go

```diff
@@ -0,0 +1,463 @@
+package csidriveroperator
+
+import (
+	"context"
+	"time"
+
+	operatorapi "github.com/openshift/api/operator/v1"
+	"github.com/openshift/cluster-storage-operator/pkg/csoclients"
+	"github.com/openshift/cluster-storage-operator/pkg/operator/csidriveroperator/csioperatorclient"
+	"github.com/openshift/cluster-storage-operator/pkg/operatorclient"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+	"k8s.io/apimachinery/pkg/api/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+	appslisters "k8s.io/client-go/listers/apps/v1"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/klog/v2"
+)
+
+// This OLMOperatorRemovalController deletes pre-existing CSI driver installed by
+// OLM to new CSI driver namespace.
+// Steps performed:
+// 1. Remove old OLM Subscription and CSV. Remember the operator namespace
+//    (in Storage CR annotation), just in case the controller is restarted
+//    after Subscription removal.
+// 2. Remove the CSI driver Deployment and DaemonSet and wait until they're
+//    really removed. Keep the old namespace around, as it can contain
+//    Manila secrets that are used by existing PVs!
+// 3. Remove the old CR (incl. removing of all of its finalizers).
+// It produces following conditions:
+// <CSI driver name>OLMOperatorRemovalProgressing/Degraded: for status reporting.
+// <CSI driver name>OLMOperatorRemovalAvailable: to signal that the removal has been complete
+type OLMOperatorRemovalController struct {
+	name             string
+	operatorClient   *operatorclient.OperatorClient
+	olmOptions       *csioperatorclient.OLMOptions
+	dynamicClient    dynamic.Interface
+	kubeClient       kubernetes.Interface
+	nsLister         corelisters.NamespaceLister
+	deploymentLister appslisters.DeploymentLister
+	daemonSetLister  appslisters.DaemonSetLister
+	eventRecorder    events.Recorder
+	factory          *factory.Factory
+
+	olmOperatorNamespace string
+}
+
+const (
+	olmOperatorRemovalControllerName = "OLMOperatorRemoval"
+
+	olmOperatorNamespaceAnnotation = ".olm-removal.storage.openshift.io/namespace"
+	olmSource                      = "redhat-operators"
+	olmSourceNamespace             = "openshift-marketplace"
+
+	oldCRName = "cluster"
+)
+
+func NewOLMOperatorRemovalController(
+	csiOperatorConfig csioperatorclient.CSIOperatorConfig,
+	clients *csoclients.Clients,
+	eventRecorder events.Recorder,
+	resyncInterval time.Duration,
```

Comment:
Should this controller have its own resync interval?

Reply:
Not sure why. Do you think it should resync more often?

Reply:
Say an error happens during sync, the controller will report a degraded condition and wait 20 (for example) minutes to resync; in the meantime, the DeploymentController will wait to deploy the CSI Driver operator...

If this is correct, IMO we should special-case this controller and hardcode a faster (but not much) resync interval... does this makes sense?

Reply:
BTW, this can be important if we'll rely on resyncs like here: https://github.com/openshift/cluster-storage-operator/pull/66#discussion_r463617363